cupy/cusolver.py

```diff
@@ -532,3 +532,135 @@ def gesv(a, b):
+        nrhs = 1
+    elif b.ndim == 2:
+        nrhs = b.shape[1]
+    else:
+        raise ValueError('b.ndim must be 1 or 2 (actual: {})'.format(b.ndim))
+    if a.shape[0] != b.shape[0]:
+        raise ValueError('shape mismatch (a: {}, b: {}).'.
+                         format(a.shape, b.shape))
+    if a.dtype != b.dtype:
+        raise ValueError('dtype mismatch (a: {}, b: {}).'.
+                         format(a.dtype, b.dtype))
+
+    dtype = a.dtype
+    if dtype == 'f':
+        t = 's'
+    elif dtype == 'd':
+        t = 'd'
+    elif dtype == 'F':
+        t = 'c'
+    elif dtype == 'D':
+        t = 'z'
+    else:
+        raise ValueError('unsupported dtype (actual: {})'.format(dtype))
+
+    geqrf_helper = getattr(_cusolver, t + 'geqrf_bufferSize')
+    geqrf = getattr(_cusolver, t + 'geqrf')
+    trsm = getattr(_cublas, t + 'trsm')
+    if t in 'sd':
+        ormqr_helper = getattr(_cusolver, t + 'ormqr_bufferSize')
+        ormqr = getattr(_cusolver, t + 'ormqr')
+    else:
+        ormqr_helper = getattr(_cusolver, t + 'unmqr_bufferSize')
+        ormqr = getattr(_cusolver, t + 'unmqr')
+
+    no_trans = _cublas.CUBLAS_OP_N
+    if dtype.char in 'fd':
+        trans = _cublas.CUBLAS_OP_T
+    else:
+        trans = _cublas.CUBLAS_OP_C
+
+    m, n = a.shape
+    mn_min = min(m, n)
+    dev_info = _cupy.empty(1, dtype=_numpy.int32)
+    tau = _cupy.empty(mn_min, dtype=dtype)
+    cusolver_handle = _device.get_cusolver_handle()
+    cublas_handle = _device.get_cublas_handle()
+    a = a.copy(order='F')
+    b = b.copy(order='F')
+
+    if m >= n:  # over/well-determined systems
+        # geqrf (QR decomposition, A = Q * R)
+        ws_size = geqrf_helper(cusolver_handle, m, n, a.data.ptr, m)
+        workspace = _cupy.empty(ws_size, dtype=dtype)
+        geqrf(cusolver_handle, m, n, a.data.ptr, m, tau.data.ptr,
+              workspace.data.ptr, ws_size, dev_info.data.ptr)
+        _cupy.linalg.util._check_cusolver_dev_info_if_synchronization_allowed(
+            geqrf, dev_info)
+
+        # ormqr (Computes Q^T * B)
+        ws_size = ormqr_helper(
+            cusolver_handle, _cublas.CUBLAS_SIDE_LEFT, trans, m, nrhs, mn_min,
+            a.data.ptr, m, tau.data.ptr, b.data.ptr, m)
+        workspace = _cupy.empty(ws_size, dtype=dtype)
+        ormqr(cusolver_handle, _cublas.CUBLAS_SIDE_LEFT, trans, m, nrhs,
+              mn_min, a.data.ptr, m, tau.data.ptr, b.data.ptr, m,
+              workspace.data.ptr, ws_size, dev_info.data.ptr)
+        _cupy.linalg.util._check_cusolver_dev_info_if_synchronization_allowed(
+            ormqr, dev_info)
+
+        # trsm (Solves R * X = (Q^T * B))
+        trsm(cublas_handle, _cublas.CUBLAS_SIDE_LEFT,
+             _cublas.CUBLAS_FILL_MODE_UPPER, no_trans,
+             _cublas.CUBLAS_DIAG_NON_UNIT, mn_min, nrhs, 1, a.data.ptr, m,
+             b.data.ptr, m)
+
+        return b[:n]
+
+    else:  # under-determined systems
+        a = _cupy.asfortranarray(a.conj().T)
```

Comment:
Note: Use `a.T.conj(order='F')` after `cupy.ndarray.conj` method supports the `order` argument.

Reply:
It seems `cupy.ndarray.conj` does not support the argument `order`...

```
>>> import cupy
>>> a = cupy.random.random((4, 5))
>>> b = a.T.conj(order='F')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: conj() takes no keyword arguments
```